libs/basic_execution/include/hpx/basic_execution/this_thread.hpp

```diff
@@ -20,47 +20,66 @@
 #include <cstddef>
 #include <cstdint>
 
-namespace hpx { namespace basic_execution { namespace this_thread {
+namespace hpx { namespace basic_execution {
 
-    namespace detail {
+    ///////////////////////////////////////////////////////////////////////////
+    // types of OS threads registered with the runtime
+    enum class thread_type
+    {
+        unknown = -1,
+        main_thread = 0,
+        worker_thread,
+        io_thread,
+        timer_thread,
+        parcel_thread,
+        custom_thread
+    };
```

Comment:
I don't think this belongs here. This is purely a runtime concept and could live in e.g. the thread mapper files. Having `hpx::basic_execution` in the thread mapper API is a bit odd as well (in light of defining our public API; I'm aware that we have other existing "public" APIs that use the nested module namespaces, but that was from before our current API discussions).

Reply:
These are needed by the `thread_executors` module (amongst other places). What module would you suggest I move that `enum` to instead?

Reply:
Also, what namespace would you suggest we move the thread mapper to?

Reply:
Ah, right... Let's leave it here for now in that case, but since the thread executors are deprecated I would move this up to the `runtime_local` module from #4606 when we remove the thread executors. Or am I missing some other non-runtime place where the enum is used?

I would put `thread_mapper` in `hpx::runtime_local` (the `thread_mapper` files are already in that module in #4606). If this is really intended to be in the public API we should probably have free functions for the rest of the thread mapper member functions (just like `thread_mapper::(un)register_thread` already have `hpx::(un)register_thread`). If we're unsure let's not have the wrappers until we know better.

Reply:
I have left the enum in the `basic_execution` module for now. However, it's now in a separate file and can easily be moved. I have also added an API function that allows to access the data that is stored by the runtime for each registered thread: `hpx::get_os_thread_data()`